graphula-core/src/Graphula.hs

```diff
@@ -429,85 +461,230 @@ class HasDependencies a where
+    , PersistEntityBackend a ~ SqlBackend
+    , PersistEntity a
+    , Typeable a
+    )
+
 
 {-|
-  Generate and edit a value with data dependencies. This leverages
+  Generate a value with data dependencies. This leverages
   'HasDependencies' to insert the specified data in the generated value. All
-  dependency data is inserted after editing operations.
+  dependency data is inserted after any editing operations.
 
-  > nodeEdit @Dog (ownerId, veterinarianId) $ \dog ->
+  > node @Dog (ownerId, veterinarianId) mempty
+  > node @Dog (ownerId, veterinarianId) $ edit $ \dog ->
   >   dog {name = "fido"}
--}
-nodeEditWith
-  :: forall a m
-   . GraphulaContext m '[a]
-  => Dependencies a
-  -> (a -> a)
-  -> m (Entity a)
-nodeEditWith dependencies edits = 10 `attemptsToInsertWith` do
-  x <- (`dependsOn` dependencies) . edits <$> generateNode
-  logNode x
-  pure x
 
-{-|
-  Generate a value with data dependencies. This leverages 'HasDependencies' to
-  insert the specified data in the generated value.
+  A value that has an externally managed key will always require an extra argument
 
-  > nodeEdit @Dog (ownerId, veterinarianId)
+  > someKey <- generateKey
+  > node @Cat (ownerId, veterinarianId) someKey mempty
+  > anotherKey <- generateKey
+  > node @Cat (ownerId, veterinarianId) anotherKey $ edit $ \cat ->
+  >   cat {name = "milo"}
 -}
-nodeWith
-  :: forall a m . GraphulaContext m '[a] => Dependencies a -> m (Entity a)
-nodeWith = flip nodeEditWith id
+node
+  :: forall a s m r c
+  . ( GraphulaContext m '[a]
+    , KeySource a ~ s
+    , ResolveArgs a s
+    , Args a s m ~ (Dict s c -> r)
+    , c
+    )
+  => Dependencies a
+  -> r
+node deps = resolveArgs @a @s @m deps Dict
 
-{-|
-  Generate and edit a value that does not have any dependencies.
+-- | Modify the node after it's been generated
+--
+-- @
+-- a <- node @A () $ edit $ \a -> a { someField = True }
+-- @
+--
+edit :: (a -> a) -> NodeOptions a
+edit f = mempty { nodeOptionsEdit = Kendo $ Just . f }
 
-  > nodeEdit @Dog $ \dog -> dog {name = "fido"}
--}
-nodeEdit
-  :: forall a m
-   . (GraphulaContext m '[a], Dependencies a ~ ())
-  => (a -> a)
-  -> m (Entity a)
-nodeEdit = nodeEditWith ()
+-- | Require a node to satisfy the specified predicate
+--
+-- @
+-- a <- node @A () $ suchThat $ (== True) . someField
+-- @
+--
+suchThat :: (a -> Bool) -> NodeOptions a
+suchThat f = mempty { nodeOptionsEdit = Kendo $ \a -> a <$ guard (f a) }
 
--- | Generate a value that does not have any dependencies
+-- | Options for generating an individual node
 --
--- > node @Dog
-node
-  :: forall a m . (GraphulaContext m '[a], Dependencies a ~ ()) => m (Entity a)
-node = nodeWith ()
+-- @'NodeOptions'@ can be created and combined with the Monoidal
+-- operations @'(<>)'@ and @'mempty'@.
+--
+-- @
+-- a1 <- node @A () mempty
+-- a2 <- node @A () $ edit $ \a -> a { someField = True }
+-- a3 <- node @A () $ suchThat $ (== True) . someField
+-- @
+--
+newtype NodeOptions a = NodeOptions
```

Comment:
I still kept `NodeOptions` for `edit`/`suchThat`, and now we have an extension point for other optional bits.